src/microbe_stage/MembraneShapeGenerator.cs

```diff
@@ -205,44 +210,162 @@ private static Vector2 FindClosestOrganelleHex(Vector2[] hexPositions, int hexCo
+        {
+            center += new Vector3(point.X, 0.0f, point.Y);
+        }
+
+        center /= vertexCount;
+
+        int layerCount = MembraneVerticalResolution * 2 + 1;
+
+        // The index list is actually a triangle list (each three consequtive indexes building a triangle)
+        // with the size of indexSize
+        var bufferSize = layerCount * vertexCount + 2;
+        var indexSize = (vertexCount * 2 + (layerCount - 1) * vertexCount * 2) * 3;
 
         var arrays = new Array();
         arrays.Resize((int)Mesh.ArrayType.Max);
 
         // Build vertex, index, and uv lists
 
-        // Index mapping to build all triangles
-        var indices = new int[indexSize];
-        int currentVertexIndex = 1;
+        // Write mesh data //
+        // for a point in points[,] array, on the layer of l and with original id of i,
+        // id is equal to i + l * count, where count is the amount of points on one layer
+        // (= amount of the outline points).
+        // Last two vertices are reserved for the topmost and the bottommost vertices of the mesh, respectively
+        var vertices = new Vector3[bufferSize];
+        var uvs = new Vector2[bufferSize];
+        var normals = new Vector3[bufferSize];
+
+        const float sideRounding = Constants.MEMBRANE_SIDE_ROUNDING;
+        const float smoothingPower = Constants.MEMBRANE_SMOOTHING_POWER;
+
+        float roundingMaximum = MathF.Pow(1.0f + sideRounding * 1.05f, smoothingPower);
+        float roundingMinimum = 1.0f;
 
-        for (int i = 0; i < indexSize; i += 3)
+        // Place prism points, already with squishification
+        for (int layer = 0; layer < layerCount; ++layer)
         {
-            indices[i] = 0;
-            indices[i + 1] = currentVertexIndex + 1;
-            indices[i + 2] = currentVertexIndex;
+            float widthModifier = 1.0f - (MathF.Pow(sideRounding * MathF.Abs(layer - MembraneVerticalResolution)
+                / MembraneVerticalResolution + 1.0f, smoothingPower) - roundingMinimum) / roundingMaximum;
+
+            float vertical = height * (layer - MembraneVerticalResolution) / MembraneVerticalResolution;
+
+            // Iterate through outline points
+            for (int i = 0; i < vertexCount; ++i)
+            {
+                Vector3 point = new Vector3(vertices2D[i].X, 0.0f, vertices2D[i].Y);
+
+                Vector3 offsetFromCenter = point - center;
+                offsetFromCenter *= widthModifier;
 
-            ++currentVertexIndex;
+                point = new Vector3(center.X + offsetFromCenter.X, vertical, center.Z + offsetFromCenter.Z);
+
+                vertices[i + layer * vertexCount] = point;
+            }
         }
 
-        // Write mesh data //
-        var vertices = new Vector3[bufferSize];
-        var uvs = new Vector2[bufferSize];
+        vertices[vertices.Length - 1] = new Vector3(center.X,
+            vertices[(layerCount - 1) * vertexCount].Y, center.Z); // Top vertex
```

Comment:
The usual style in thrive is to not use trailing comments on lines, so these comments could be moved to be before the lines they comment on for style uniformity.